File: Sliding Window/Length of the longest substring.py
'''
https://www.geeksforgeeks.org/problems/length-of-the-longest-substring3036/1
'''


class Solution:
    def longestUniqueSubsttr(self, S):
        length = len(S)
        j = 0
        i = 0
        result = 0
        space = {}
        ans = 0
        while j<length:
            space[S[j]] = space.get(S[j],0)+1
            while i<=j and space[S[j]]!=1:
                space[S[i]]-=1
                if space[S[i]]==0:
                    del space[S[i]]
                i+=1
            ans = max(ans,j-i+1)
            j+=1
        return ans

File: Sliding Window/Length of the longest substring.cpp
/*
https://www.geeksforgeeks.org/problems/length-of-the-longest-substring3036/1
  */
class Solution{
    public:
    int longestUniqueSubsttr(string S){
        int result=0;
        int left=0;
        vector<int> space(26,0);
        for(int right=0;right<S.size();right++){
            int index = S[right]-'a';
            space[index]+=1;
            while(space[index]>1){
                int tempIndex = S[left]-'a';
                space[tempIndex]-=1;
                left++;
            }
            result = max(result,right-left+1);
        }
        return result;
    }
};
